Custom errors, extending Error (#163)
javascript-tutorial/it.javascript.info: 1-js/10-error-handling/2-custom-errors/article.md

Man mano che l'applicazione cresce, i nostri errori formeranno naturalmente una gerarchia. Per esempio, `HttpTimeoutError` può ereditare da `HttpError`, e così via.

## Estendere "Error"

Come esempio, consideriamo una funzione `readUser(json)` che dovrebbe leggere un JSON con i dati dell'utente.

Questo è un esempio di come dovrebbe apparire un `json` valido:
```js
let json = `{ "name": "John", "age": 30 }`;
```

Internamente, useremo `JSON.parse` che, se riceve un `json` malformato, lancia `SyntaxError`. Ma anche se il `json` è sintatticamente corretto, questo non significa che sia un utente valido, ok? Potrebbero mancare i dati necessari. Ad esempio, potrebbe non avere le proprietà `name` e `age` che sono essenziali per i nostri utenti.

La nostra funzione `readUser(json)` non solo leggerà il JSON, ma validerà i dati. Se non ci sono i campi richiesti, o il formato è errato, allora c'è un errore. E non è un `SyntaxError`, dato che è sintatticamente corretto, ma un altro tipo di errore. Lo chiameremo `ValidationError` e creeremo una classe per esso. Un errore di questo tipo dovrebbe contenere le informazioni riguardo il campo incriminato.

> La nostra classe `ValidationError` dovrebbe ereditare dalla classe incorporata `Error`.

Reviewer: Qui ed in seguito "built-in class" viene tradotto come "classe incorporata". Valuta se non sarebbe meglio usare "classe integrata" , o addirittura evitare di tradurre built-in

}
```

Poniamo attenzione: alla linea `(1)` richiamiamo il construtto genitore. JavaScript ci richiede di richiamare `super` nel construtto figlio, quindi è obbligatorio. Il construtto genitore imposta la proprietà `message`.

> Il genitore imposta anche la proprietà `name` in `"Error"`, allora nella linea `(2)` re-impostiamo il corretto valore.

Reviewer: ```suggestion
Il genitore imposta anche la proprietà `name` in `"Error"`, quindi nella linea `(2)` re-impostiamo il corretto valore.
```

> Quindi proviamo ad usarlo in `readUser(json)`:

Reviewer: ```suggestion
Proviamo ad usarlo in `readUser(json)`:
```

  }
}

// Utilizzo
function readUser(json) {
  let user = JSON.parse(json);

  if (!user.age) {
    throw new ValidationError("Manca un campo: age");
  }
  if (!user.name) {
    throw new ValidationError("Manca un campo: name");
  }

  return user;
}

// Esempio funzionante con try..catch

try {
  let user = readUser('{ "age": 25 }');
} catch (err) {
  if (err instanceof ValidationError) {
*!*
    alert("Invalid data: " + err.message); // Dati non Validi: Manca un campo: name
*/!*
  } else if (err instanceof SyntaxError) { // (*)
    alert("JSON Syntax Error: " + err.message);
  } else {
    throw err; // errore sconosciuto, lo rilancio (**)
  }
}
```

Il blocco `try..catch` nel codice qui sopra gestisce sia il nostro `ValidationError` che l'errore `SyntaxError` lanciato da `JSON.parse`.

Poniamo particolare attenzione a come usiamo `instanceof` per verificare errori specifici nella linea `(*)`.

> Noi potremmo anche verificare tramite `err.name`, tipo così:

Reviewer: ```suggestion
Noi potremmo anche verificare tramite `err.name`, tipo così:
```
```suggestion
Potremmo anche verificare tramite `err.name`, nel seguente modo:
```

```js
// ...
// al posto di (err instanceof SyntaxError)
} else if (err.name == "SyntaxError") { // (*)
// ...
```  

> La versione con `instanceof` è sicuramente migliore, perchè in futuro andremo a estendere `ValidationError`, creando sottotipi di esso, come `PropertyRequiredError`. E il controllo `instanceof` continuerà a funzionare per le nuovi classi ereditate. Quindi è a prova di futuro.

Reviewer: ```suggestion
La versione con `instanceof` è sicuramente migliore, perchè in futuro andremo a estendere `ValidationError`, creando sottotipi di esso, come `PropertyRequiredError`. E il controllo `instanceof` continuerà a funzionare per le nuove classi ereditate. Quindi è a prova di futuro.
```

}
```

La nuova classe `PropertyRequiredError` è facile da usare: dobbiamo solamente fornire il nome della proprietà: `new PropertyRequiredError(property)`. Il messaggio `message` è generato dal construttore.

> Poniamo particolare attenzione al fatto che `this.name` nel construttore `PropertyRequiredError` è di nuovo assegnato manualmente. Questa cosa potrebbe risultare un po noiosa -- assegnare `this.name = <class name>` in ogni errore personalizzato. Possiamo evitarlo creando la nostra classe "basic error" che assegna `this.name = this.constructor.name`. E poi ereditare tutti i nostri errori personalizzati da esso.

Reviewer: ```suggestion
Poniamo particolare attenzione al fatto che `this.name` nel construttore `PropertyRequiredError` è di nuovo assegnato manualmente. Questa cosa potrebbe risultare un po' noiosa -- assegnare `this.name = <class name>` in ogni errore personalizzato. Possiamo evitarlo creando la nostra classe "basic error" che assegna `this.name = this.constructor.name`, quindi ereditare da questa  tutti i nostri errori personalizzati.
```

alert( new PropertyRequiredError("field").name ); // PropertyRequiredError
```

Ora il codice degli errori personalizzati sarà più corto, specialmente `ValidationError`, dato che ci siamo sbarazzati della linea con `"this.name = ..."` nel construttore.

## Wrapping exceptions

Lo scopo della funzione `readUser` nel codice precedente è di "leggere i dati dell'utente". Possono accadere diverse cose durante questo processo. Per adesso abbiamo `SyntaxError` e `ValidationError`, ma in futuro la funzione `readUser` potrebbe crescere e probabilmente generare altri tipi di errore.

Il codice che richiama `readUser` dovrebbe gestire questi errori. Per ora utilizziamo diversi `if` nel blocco `catch`, che verificano la classe, ne gestiscono gli errori e rilanciano quelli sconosciuti.

> Lo schema è tipo così:

Reviewer: ```suggestion
Lo schema è simile al seguente:
```

    // gestisco gli errori di sintassi
  } else {
    throw err; // errore sconosciuto, lo rilancio
  }
}
```

Nel codice qui sopra possiamo notare due tipi di errore, ma ce ne possono essere molti di più.

Se la funzione `readUser` genera diversi tipi di errore, allora dovremmo chiederci: vogliamo veramente controllare tutti i tipi di errore un alla volta ogni volta?

Spesso la risposta è "No": vorremo stare tutto sommato "un livello sopra tutto questo". A noi interessa sapere se c'è un "errore nella lettura dei dati" -- perchè esattamente questo accada è spesso irrilevante (il messaggio di errore già lo descrive). O, ancora meglio, vorremo avere un modo per ottenere i dettagli dell'errore, ma solo quando ne abbiamo bisogno.

La tecnica che andiamo qui a descrivere è chiamata "wrapping exceptions".

1. Creeremo una nuova classe `ReadError` che rappresenta un errore generico di "lettura dei dati".
2. La funzione `readUser` catturerà gli errori di lettura che avvengono al suo interno, come `ValidationError` e `SyntaxError`, e genererà un `ReadError`.
3. L'oggetto `ReadError` terrà i riferimenti all'errore originale nella sua proprietà `cause`.

> Quindi il codice che richiama `readUser` dovrà solamente controllare se si verifica un `ReadError`, e non ogni tipo di errore nella lettura dei dati. E se abbiamo la necessità di approfondire riguardo un errore, lo potremo fare controllando la proprietà `cause`.

Reviewer: ```suggestion
Quindi il codice che richiama `readUser` dovrà solamente controllare se si verifica un `ReadError`, e non ogni tipo di errore nella lettura dei dati. E se abbiamo la necessità di approfondire i dettagli dell'errore, lo potremo fare controllando la proprietà `cause`.
```

> Questo il codice che definisce `ReadError` e la dimostrazione di come usarlo in `readUser` e nel `try..catch`:

Reviewer: ```suggestion
Questo è il codice che definisce `ReadError` e la dimostrazione di come usarlo in `readUser` e nel `try..catch`:
```
